kit/Kit.cpp

```diff
@@ -1994,6 +2005,24 @@ class KitSocketPoll final : public SocketPoll
         _document = std::move(document);
     }
 
+    // unusual LOK event from another thread, push into our loop to process.
+    static bool pushToMainThread(LibreOfficeKitCallback callback, int type, const char *p, void *data)
+    {
+        if (mainPoll && mainPoll->getThreadOwner() != std::this_thread::get_id())
+        {
+            LOG_TRC("Unusual push callback to main thread");
+            std::shared_ptr<std::string> pCopy;
+            if (p)
+                pCopy = std::make_shared<std::string>(p, strlen(p));
+            mainPoll->addCallback([=]{
+                LOG_TRC("Unusual process callback in main thread");
+                callback(type, pCopy ? pCopy->c_str() : nullptr, data);
```

Comment:
Is `data` guaranteed to live beyond the callback? Could the thread end-up with a dangling pointer?

Reply:
Good catch, at least core.git hands out a data pointer which is only valid for the duration of the callback, and then core frees it. I'm not sure if there is some layer between this function and core that would perhaps avoid this problem in practice.

Reply:
Well the data is really a closure that we provided to them in the first instance =) I checked the ViewCallback which has a CallbackDescriptor pointer - which is eternal and I updated the comment there to help review ;-) For the GlobalCallback the *data is the Document and our lifecycle there is basically being killed by a remote process for fast cleanup, or calling _exit() similarly, so I think we're good here.

Reply:
Ah I see, then this is not a problem.